### PR TITLE
[hail/ptypes] precise CastRename ptype

### PR DESCRIPTION
**Pull request by akotlar**

### CODE REVIEW

File: hail/src/test/scala/is/hail/expr/ir/IRSuite.scala [hunk 1]
```diff
@@ -199,6 +199,24 @@ class IRSuite extends HailSuite {
       FastIndexedSeq(Row(1)))
   }
 
+  @Test def testCastRenameIR() {
+    var expectedPType: PType = PCanonicalStruct(true, "foo" -> PInt32(true))
```
Comment (tpoterba) [inline hail/src/test/scala/is/hail/expr/ir/IRSuite.scala]: add tests with other collection types

File: hail/src/main/scala/is/hail/expr/ir/InferPType.scala [hunk 2]
```diff
@@ -46,6 +46,23 @@ object InferPType {
     }
   }
 
+  def deepRenameStruct(typeWithNames: Type, targetPType: PType): PType = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: This should be a method on PType, with a default implementation `return this`, and collection types override (e.g. for PStruct):

```scala
override def deepRename(t: Type) = deepRenameStruct(t.asInstanceOf[TStruct])
def deepRenameStruct(t: TStruct)
```

(PCanonicalBaseStruct implements the `deepRenameStruct`)

You're also missing all the other collections.

**Comment by tpoterba** [comment]
That second comment was supposed to be in the first review, but github deleted it. Thanks, github.

File: hail/src/main/scala/is/hail/expr/ir/InferPType.scala (hunk 2 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: Sure. It's hard to understand how CastRename is used without seeing test cases. The only test cases that existed for CastRename in IRSuite were for Struct and Array.

Will address these likely after I return.

**Comment by akotlar** [comment]
Tim, decided to address this now so that you weren't blocked early next week

File: hail/src/test/scala/is/hail/expr/ir/IRSuite.scala (hunk 1 above)
Comment (akotlar) [reply hail/src/test/scala/is/hail/expr/ir/IRSuite.scala]: done

**Pull request merged**